FirebaseSegmentation/Sources/SEGDatabaseManager.m

```diff
@@ -0,0 +1,419 @@
+            SegmentationAddSkipBackupAttributeToItemAtPath(dbPath);
+          }
+        } else {
+          [strongSelf logDatabaseError];
+          completionHandler(NO, @{@"ErrorDescription" : @"Could not create database."});
+        }
+      }
+    } else {
+      [strongSelf logDatabaseError];
+      completionHandler(NO, @{@"ErrorDescription" : @"Error creating database."});
+    }
+  });
+}
+
+- (void)removeDatabase:(NSString *)path completion:(SEGRequestCompletion)completionHandler {
+  __weak SEGDatabaseManager *weakSelf = self;
+  dispatch_async(_databaseOperationQueue, ^{
+    SEGDatabaseManager *strongSelf = weakSelf;
+    if (!strongSelf) {
+      return;
+    }
+    [strongSelf removeDatabase:path];
+  });
+}
+
+#pragma mark - Private Methods
+
+- (NSDictionary *)loadMainTable {
+  NSString *SQLQuery = [NSString
+      stringWithFormat:@"SELECT %@, %@, %@, %@ FROM %@", kMainTableColumnApplicationIdentifier,
+                       kMainTableColumnCustomInstallationIdentifier,
+                       kMainTableColumnFirebaseInstallationIdentifier,
+                       kMainTableColumnAssociationStatus, kMainTableName];
+
+  sqlite3_stmt *statement = [self prepareSQL:[SQLQuery cStringUsingEncoding:NSUTF8StringEncoding]];
+  if (!statement) {
+    return nil;
+  }
+
+  NSMutableDictionary<NSString *, NSDictionary<NSString *, NSString *> *> *associations =
+      [[NSMutableDictionary alloc] init];
+  while (sqlite3_step(statement) == SQLITE_ROW) {
+    NSString *firebaseApplicationName =
+        [[NSString alloc] initWithUTF8String:(char *)sqlite3_column_text(statement, 0)];
+    NSString *customInstallationIdentifier =
+        [[NSString alloc] initWithUTF8String:(char *)sqlite3_column_text(statement, 1)];
+    NSString *firebaseInstallationIdentifier =
+        [[NSString alloc] initWithUTF8String:(char *)sqlite3_column_text(statement, 2)];
+    NSString *associationStatus =
+        [[NSString alloc] initWithUTF8String:(char *)sqlite3_column_text(statement, 3)];
+    NSDictionary<NSString *, NSString *> *associationData = @{
+      kSEGCustomInstallationIdentifierKey : customInstallationIdentifier,
+      kSEGFirebaseInstallationIdentifierKey : firebaseInstallationIdentifier,
+      kSEGAssociationStatusKey : associationStatus
+    };
+    [associations setObject:associationData forKey:firebaseApplicationName];
+  }
+  sqlite3_finalize(statement);
+  return associations;
+}
+
+/// Returns the current version of the Remote Config database.
++ (NSString *)pathForSegmentationDatabase {
+  NSArray<NSString *> *dirPaths =
+      NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory, NSUserDomainMask, YES);
+  NSString *appSupportPath = dirPaths.firstObject;
+  NSArray<NSString *> *components =
+      @[ appSupportPath, kApplicationSupportSubDirectory, kDatabaseName ];
+  return [NSString pathWithComponents:components];
+}
+
+- (BOOL)createTableSchema {
+  SEG_MUST_NOT_BE_MAIN_THREAD();
```

Comment:
Don't rely on checking`[NSThread isMainThread]` for ensuring the main thread is not blocked. For example, if a called on the main thread `dispatch_sync`s to a different queue to invoke this method, one of two things will happen:

1. GCD will [optimize](https://developer.apple.com/documentation/dispatch/1452870-dispatch_sync) and execute the block on the main thread anyway, though under the guise of a different queue
2. The main thread blocks anyway while waiting for the block to execute.

What is this main thread check trying to achieve?